azurerm/internal/services/mssql/mssql_virtual_machine_resource.go

```diff
@@ -486,15 +486,33 @@ func flattenSqlVirtualMachineStorageConfigurationSettings(input *sqlvirtualmachi
 		return []interface{}{}
 	}
 
-	return []interface{}{
-		map[string]interface{}{
-			"disk_type":             string(input.DiskConfigurationType),
-			"storage_workload_type": storageWorkloadType,
-			"data_settings":         flattenSqlVirtualMachineStorageSettings(input.SQLDataSettings),
-			"log_settings":          flattenSqlVirtualMachineStorageSettings(input.SQLLogSettings),
-			"temp_db_settings":      flattenSqlVirtualMachineStorageSettings(input.SQLTempDbSettings),
-		},
+	output := make(map[string]interface{})
+
+	if v := string(input.DiskConfigurationType); v != "" {
+		output["disk_type"] = v
+	}
+
+	if storageWorkloadType != "" {
+		output["storage_workload_type"] = storageWorkloadType
+	}
```

Comment:
this possibly wants setting as an empty string so there's always a value?

Reply:
Unfortunately the problem is that Terraform shows state drift if the containing list has any elements when there are none in the config. It looks like the API has started returning an array having zero values instead of omitting it from the response, which is why I've made it try and detect that. Is there is a better way to avoid state drift in such circumstances?

Reply:
agreed we should track an empty string in the state in all cases - but if the nested object is returned as an empty object (e.g. all fields are empty) then we'd need to do conditionally return an empty object rather than the nested object - and presumably raise an API bug for that?

Reply:
@tombuildsstuff Yep an empty object is exactly what's being returned. I'll raise an API bug today and adjust the logic so that an empty string is saved if the other attributes are present and non-zero.

Reply:
given this affects the entire block - presuming we're using intermediate variables, we should be able to conditionally check the set and conditionally return the nested object, e.g.:

```
foo := ""
if props.Foo != nil { foo = *props.Foo }

bar := ""
if props.Bar != nil { bar = *props.Bar }

if foo == "" && bar == "" {  return []interface{}{} }

return []interface{}{
  map[string]interface{}{
    "foo": foo,
    "bar": bar,
  },
}
```

(we've a few instances of this of late)

Reply:
Cool, makes sense. Have adjusted it to always set all attributes [in the block], but return an empty slice if they all have zero values.